tests: secureboot: refactor to facilitate extension

alexgg opened the pull request:
As the secure boot feature is extended to more device types, the test needs to be easy to extend.

Change-type: patch


---
### Contributor checklist
<!-- For completed items, change [ ] to [x].  -->
- [x] Changes have been tested
  - [x] Covered in automated test suite
  - [ ] Manual test case recorded
- [x] `Change-type` present on at least one commit
- [x] `Signed-off-by` is present
- [x] The PR complies with the [Open Embedded Commit Patch Message Guidelines](http://www.openembedded.org/wiki/Commit_Patch_Message_Guidelines)
<!-- optional: `Changelog-entry` present on at least one commit if you want to set the changelog entry manually-->

### Reviewer Guidelines
- When submitting a review, please pick:
  - '*Approve*' if this change would be acceptable in the codebase (even if there are minor or cosmetic tweaks that could be improved).
  - '*Request Changes*' if this change would not be acceptable in our codebase (e.g. bugs, changes that will make development harder in future, security/performance issues, etc).
  - '*Comment*' if you don't feel you have enough information to decide either way (e.g. if you have major questions, or you don't understand the context of the change sufficiently to fully review yourself, but want to make a comment)

rcooke-warwick commented on tests/suites/os/tests/secureboot/index.js:
```diff
@@ -1,145 +1,197 @@
-						this.worker.executeCommandInHostOS('modprobe pcan_netdev', this.link),
-						'Module with valid signature loads',
-					);
+	async resetWorker() {
```
nit but isn't this "resetDut" ?

alexgg replied:
This was the original name - I think `resetDut` is more descriptive, I can change it in a different commit not to confuse the refactoring here.

rcooke-warwick commented on tests/suites/os/tests/secureboot/index.js:
```diff
@@ -1,145 +1,197 @@
+		this.impl = impl;
+	}
+
+	setImplementation(impl) {
```
What is this intended to be used for?

alexgg replied:
Actually, we can just run different tests for each implementation so this will not be needed, I'll remove it.

rcooke-warwick commented on tests/suites/os/tests/secureboot/index.js:
```diff
@@ -1,145 +1,197 @@
+		{
+			title: 'check secure boot',
+			run: async function(test) {
+				const impl = new testSecureBoot(new uefiSecureBoot(test, this.worker, this.suite.context, this.os.image.path));
```
How can we know which class to make an instance of based on the device type?

alexgg replied:
We don't need to - basically the test runs all implementations and the `isSecureBootEnabled` function will only return true if the implementation is meant to run.

vipulgupta2048 replied:
Every implementation of secure boot functionality we end up supporting? -- Seems like a catch-all. Or am I understanding your statement wrong @alexgg

rcooke-warwick commented on tests/suites/os/tests/secureboot/index.js:
```diff
@@ -1,145 +1,197 @@
-					);
+	async resetWorker() {
+		await this.worker.off();
+		await this.worker.flash(this.imagePath);
```
One thing I'm still unsure about when we have to test on other types of device is to what extend we will need to manipulate the flashing process - do you forsee that we might need to pass other parameters here, other than the OS image? (this is an issue separate but related to this PR)

alexgg replied:
Unless the `flash` method has a way to determine what device it is running on to decide how to program we might need to specialize the `resetWorker()/resetDUT()` function.

rcooke-warwick commented on tests/suites/os/tests/secureboot/index.js:
```diff
@@ -1,145 +1,197 @@
-			run: async function(test) {
-				const securebootEfiVarPath = '/sys/firmware/efi/efivars/SecureBoot-*';
+const securebootEfiVarPath = '/sys/firmware/efi/efivars/SecureBoot-*';
+const Bluebird = require('bluebird');
```
This doesn't seem to be used anywhere

vipulgupta2048 replied:
@alexgg Can you post what error were you getting: https://github.com/balena-os/leviathan/blob/c15d5183a61e4ed9d437642560c2db071b1d7672/core/lib/common/worker.js#L218

I wrote fetchSerial and there is bluebird already imported there?

alexgg replied:
In does seem its not needed - probably some artifact of running it in the middle of refactoring, I'll remove it.

jakogut commented on tests/suites/os/tests/secureboot/index.js:
```diff
@@ -1,145 +1,193 @@
+			this.waitForSerialOutput(/bootx64.efi: Access Denied/),
+			'Firmware will not load bootloader that fails verification',
+		)
+		).then(this.resetWorker());
```
@alexgg This is broken. `.then()` accepts a callback function for when the promise completes. The original code is written as `.then(resetWorker)`, which means "execute resetWorker and return the resulting promise". This code executes the function immediately, and the resulting promise is not a callable handler for the previous promise resolving.

jakogut replied:
I suspect the effect this has is resetting the DUT as soon as the statement is evaluated, rather than after the test is completed.

alexgg replied:
indeed @jakogut thanks - I changed it to using an arrow function

vipulgupta2048 commented on tests/suites/os/tests/secureboot/index.js:
```diff
@@ -1,145 +1,193 @@
+	async waitForSerialOutput(pattern, slice=0, retries=60, delay=1000) {
+		await this.utils.waitUntil(
+			async () => this.worker.fetchSerial().then(
+				serialLogs => serialLogs.split('\n')
+				.slice(slice)
+				.join('\n')
+				.match(pattern)
+			), false, retries, delay
+		);
+	}
```
```suggestion
				(serialLogs) => {
				  serialLogs.split('\n')
				  .slice(slice)
				  .join('\n')
				  .match(pattern)
				}
			), false, retries, delay
		);
	}
```

NIT: Improving readability

vipulgupta2048 commented on tests/suites/os/tests/secureboot/index.js:
```diff
@@ -1,145 +1,193 @@
+	title: 'secure boot tests',
+	tests: [
+		{
+			title: 'check UEFI secure boot',
+			title: 'check secure boot',
```
TAP JSON won't be happy accepting 2 keys of the same property `title`? Remove one of them

alexgg commented:
@resin-jenkins retest this please

alexgg commented:
@resin-jenkins retest this please

alexgg commented:
Re-tried `qemu-generic-amd64-sb`

alexgg commented:
@resin-jenkins retest this please

alexgg commented:
@resin-jenkins retest this please

rcooke-warwick commented:
latest commit fixes cloud suite

alexgg commented:
@resin-jenkins retest this please

alexgg commented:
@resin-jenkins retest this please

alexgg commented:
@resin-jenkins retest this please

alexgg commented:
@resin-jenkins retest this please

alexgg commented:
@resin-jenkins retest this please

alexgg commented:
@resin-jenkins retest this please

alexgg commented:
@resin-jenkins retest this please

alexgg commented:
@resin-jenkins retest this please

alexgg commented:
@resin-jenkins retest this please

alexgg commented:
```
jenkins-leviathan-v2-template-55678-client-1  | [2024-07-05T20:50:39.141Z][worker-os]             not ok 7 - Firmware will not load bootloader that fails verification
jenkins-leviathan-v2-template-55678-client-1  | [2024-07-05T20:50:39.141Z][worker-os]               ---
jenkins-leviathan-v2-template-55678-client-1  | [2024-07-05T20:50:39.143Z][worker-os]               at:
jenkins-leviathan-v2-template-55678-client-1  | [2024-07-05T20:50:39.144Z][worker-os]                 line: 150
jenkins-leviathan-v2-template-55678-client-1  | [2024-07-05T20:50:39.145Z][worker-os]                 column: 26
jenkins-leviathan-v2-template-55678-client-1  | [2024-07-05T20:50:39.145Z][worker-os]                 file: /data/suite/tests/secureboot/index.js
jenkins-leviathan-v2-template-55678-client-1  | [2024-07-05T20:50:39.145Z][worker-os]               found:
jenkins-leviathan-v2-template-55678-client-1  | [2024-07-05T20:50:39.145Z][worker-os]                 !error
jenkins-leviathan-v2-template-55678-client-1  | [2024-07-05T20:50:39.145Z][worker-os]                 name: Error
jenkins-leviathan-v2-template-55678-client-1  | [2024-07-05T20:50:39.145Z][worker-os]                 message: |-
jenkins-leviathan-v2-template-55678-client-1  | [2024-07-05T20:50:39.145Z][worker-os]                   Condition async () => this.worker.fetchSerial().then(
jenkins-leviathan-v2-template-55678-client-1  | [2024-07-05T20:50:39.146Z][worker-os]                   				serialLogs => {
jenkins-leviathan-v2-template-55678-client-1  | [2024-07-05T20:50:39.146Z][worker-os]                   					serialLogs.split('\n')
jenkins-leviathan-v2-template-55678-client-1  | [2024-07-05T20:50:39.146Z][worker-os]                   					.slice(slice)
jenkins-leviathan-v2-template-55678-client-1  | [2024-07-05T20:50:39.146Z][worker-os]                   					.join('\n')
jenkins-leviathan-v2-template-55678-client-1  |                   					.match(pattern)
jenkins-leviathan-v2-template-55678-client-1  | [2024-07-05T20:50:39.146Z][worker-os]                   				}) timed out
jenkins-leviathan-v2-template-55678-client-1  | [2024-07-05T20:50:39.147Z][worker-os]                 stack: >-
jenkins-leviathan-v2-template-55678-client-1  | [2024-07-05T20:50:39.147Z][worker-os]                   Error: Condition async () => this.worker.fetchSerial().then(
jenkins-leviathan-v2-template-55678-client-1  | [2024-07-05T20:50:39.147Z][worker-os]                   				serialLogs => {
jenkins-leviathan-v2-template-55678-client-1  | [2024-07-05T20:50:39.147Z][worker-os]                   					serialLogs.split('\n')
jenkins-leviathan-v2-template-55678-client-1  | [2024-07-05T20:50:39.147Z][worker-os]                   					.slice(slice)
jenkins-leviathan-v2-template-55678-client-1  | [2024-07-05T20:50:39.148Z][worker-os]                   					.join('\n')
jenkins-leviathan-v2-template-55678-client-1  | [2024-07-05T20:50:39.148Z][worker-os]                   					.match(pattern)
jenkins-leviathan-v2-template-55678-client-1  | [2024-07-05T20:50:39.148Z][worker-os]                   				}) timed out
jenkins-leviathan-v2-template-55678-client-1  | [2024-07-05T20:50:39.148Z][worker-os]                       at _waitUntil (/usr/app/lib/common/utils.js:103:11)
jenkins-leviathan-v2-template-55678-client-1  | [2024-07-05T20:50:39.148Z][worker-os]                       at _waitUntil (/usr/app/lib/common/utils.js:118:11)
jenkins-leviathan-v2-template-55678-client-1  |                       at Object.waitUntil (/usr/app/lib/common/utils.js:121:3)
jenkins-leviathan-v2-template-55678-client-1  | [2024-07-05T20:50:39.149Z][worker-os]                       at uefiSecureBoot.waitForSerialOutput (/data/suite/tests/secureboot/index.js:28:3)
jenkins-leviathan-v2-template-55678-client-1  | [2024-07-05T20:50:39.149Z][worker-os]               source: |
jenkins-leviathan-v2-template-55678-client-1  | [2024-07-05T20:50:39.149Z][worker-os]                 		).then(() => this.worker.executeCommandInHostOS('reboot', this.link)
jenkins-leviathan-v2-template-55678-client-1  | [2024-07-05T20:50:39.149Z][worker-os]                 		).then(() => this.test.resolves(
jenkins-leviathan-v2-template-55678-client-1  | [2024-07-05T20:50:39.149Z][worker-os]                 -------------------------^
jenkins-leviathan-v2-template-55678-client-1  |                 			this.waitForSerialOutput(/bootx64.efi: Access Denied/),
jenkins-leviathan-v2-template-55678-client-1  | [2024-07-05T20:50:39.150Z][worker-os]                 			'Firmware will not load bootloader that fails verification',
jenkins-leviathan-v2-template-55678-client-1  | [2024-07-05T20:50:39.150Z][worker-os]               ...
jenkins-leviathan-v2-template-55678-client-1  | [2024-07-05T20:50:39.165Z][worker-os] 
jenkins-leviathan-v2-template-55678-client-1  | [2024-07-05T20:50:39.165Z][worker-os]             Bail out! Firmware will not load bootloader that fails verification
```

alexgg commented:
@resin-jenkins retest this please

alexgg commented:
@resin-jenkins retest this please

alexgg commented:
lgtm
